### PR TITLE
scripts: example lua script to do Rate limiting

### PR DESCRIPTION
**Pull request by Gokul-Radhakrishnan**
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by patrick-stephens** [comment]
Can you update the earlier commits as well? 

@tarruda what do you think to the scripts?

**Comment by Gokul-Radhakrishnan** [comment]
@patrick-stephens I have updated the commit messages

**Comment by tarruda** [comment]
LGTM

File: scripts/rate_limit.lua [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+end
+
+function rate_limit(tag, timestamp, record)
+    local t = os.time()
+    local current_time = get_current_time(t)
+    if current_time ~= time then
+        time = current_time
+        counter = {} -- reset the counter
+    end
+    local logsCount = counter[record["kubernetes"][key]]
```
[the remark this answers is not in the document]
Comment (Gokul-Radhakrishnan) [reply scripts/rate_limit.lua]: @tarruda Thanks for your inputs. I have updated the script with above mentioned changes

**Review by tarruda — DISMISSED** [review]

**Comment by Gokul-Radhakrishnan** [comment]
@leonardo-albertovich  @edsiper  @koleini  @fujimotos 
Can you please review this pull request?

**Comment by leonardo-albertovich** [comment]
I know virtually nothing about lua and especially how is lua embedded in fluent-bit. 

Considering that this is executed in a filter it means it will run in the main thread and concurrency wouldn't be an issue at the moment.

The only things I think could be improved are the environment variable names (a bit too generic and non descriptive) and when it comes to time keeping and period measurement I would like to know : 

1. How expensive is having that `os.time` in there running for each entry.
2. Is that a monotonic clock (doens't seem to be the case).
3. Is this clock / timing functionality something we would like to expose from fluent-bit?

**Comment by leonardo-albertovich** [comment]
Thank you very much for addressing my comments about the naming issues. Considering that @tarruda already approved the script in a comment I'll approve it formally so the tests can run (although there are no tests for the script itself) and I think @edsiper could merge it afterwards.

**Comment by Gokul-Radhakrishnan** [comment]
> The only things I think could be improved are the environment variable names (a bit too generic and non descriptive) 

Thanks for your suggestion. I have changed the variable names.
> 1. How expensive is having that `os.time` in there running for each entry.

`os.time`  is just a light wrapper around time() C stdlib function so it is cheap 
> 2. Is that a monotonic clock (doens't seem to be the case).

It is monotonic clock and it returns Unix epoch time. The Unix epoch  is the number of seconds that have elapsed since January 1, 1970 (midnight UTC). 

> 3. Is this clock / timing functionality something we would like to expose from fluent-bit?

No we are not exposing this timing functionality from fluent-bit. We are just using the timing functionality provided by lua.

**Comment by Gokul-Radhakrishnan** [comment]
@edsiper Can you please review this pull request?

**Comment by edsiper** [comment]
hi @Gokul-Radhakrishnan , thanks for your contribution!

Before merging it, would you please add:

- example configuration file in `conf/`  (adding some comments about environment vars needed)
- I see you need environment variables, please make sure the script have default values if those are not set

**Comment by Gokul-Radhakrishnan** [comment]
@edsiper Thanks for reviewing. It is not mandatory to use environment variables here. Can I go ahead and hard code the values inside the script? I feel that will be simple and more clear to the users. What is your thoughts on this? 

```
local group_key = "docker_id" -- Used to group logs. Groups are rate limited independently. 
local group_bucket_period_s = 60 -- This is the period of of time in seconds over which group_bucket_limit applies.
local group_bucket_limit = 6000 -- Maximum number logs allowed per groups over the period of group_bucket_period_s.
```

**Comment by Gokul-Radhakrishnan** [comment]
@edsiper I have added example config file

**Pull request merged**